File: asp.net-store-project.client/src/shared/StoreObject/OrderInfo.ts
import { DeliveryMethod } from "../StoreEnum/DeliveryMethod";
import { ProductInfo } from "./ProductInfo";

// Client-side OrderInfo equivalent
export type OrderInfo = {
    id: string;
    supplierId: string;
    supplierName: string;
    productsCost: number;
    transportCost: number;
    deliveryMethod: DeliveryMethod,
    products: ProductInfo[];
    customerDetails: CustomerInfo;
    adressDetails: AdressInfo;
    stageHistory: OrderStageInfo[];
};

// Client-side CustomerInfo equivalent
export type CustomerInfo = {
    name: string;
    surname: string;
    phoneNumber: string;
    email: string;
}

// Client-side AdressInfo equivalent
export type AdressInfo = {
    region: string;
    city: string;
    postalCode: string;
    streetName: string;
    houseNumber: string;
    apartmentNumber?: string;
}

// Client-side OrderStageInfo equivalent
export type OrderStageInfo = {
    type: string;
    dateOfCreation: string;
    timeStamp: string;
}

File: ASP.NET-store-project.Server/Models/StructuredData/OrderInfo.cs
namespace ASP.NET_store_project.Server.Models.StructuredData;

using DMethod = Data.Enums.DeliveryMethod; // Name collision within the class
using System.Text.Json.Serialization;
using Microsoft.OpenApi.Extensions;

// Uniform class for server-server and server-client communication
[method: JsonConstructor]
public record OrderInfo(
    string? Id, 
    string? SupplierId,
    string? SupplierName,
    decimal ProductsCost,
    decimal DeliveryCost,
    string DeliveryMethod,
    IEnumerable<ProductInfo> Products, 
    CustomerInfo CustomerDetails, 
    AdressInfo AdressDetails, 
    OrderStageInfo[] StageHistory)
{
    // Constructor for the external APIs to fill adequate attributes
    public OrderInfo(IEnumerable<ProductInfo> Products, decimal DeliveryCost, CustomerInfo CustomerDetails, AdressInfo AdressDetails)
        : this(null, null, null, 0, DeliveryCost, DMethod.Standard.GetDisplayName(), Products, CustomerDetails, AdressDetails, []) { }

}
